### ISSUE TITLE
Variable Sizer

### ISSUE THREAD

**Issue by JulienNic0las** [mementum/backtrader#72]
Hi,

I think it would be great if by default a kind of variable sizer (even a simple one, if simple it can be) could be present in backtrader. As far as I understand, sizers is a big piece of work in perspective.

Right now it is only possible to call buy or sell orders specifying a fix amount of shares (buy X shares or sell Y shares), but a common practice is to call an order for a given amount of money (buy for X' dollars). With a fix size, the strategy can ask to the broker to buy X shares, but when it comes to the broker to accomplish the order the price may have increased. As a consequence, the available cash may not be enough to buy the required X shares and the order is dismissed/rejected.

Is  such implementation feasible? If yes, is it planned in a near future, or will it be implemented along with more other complex sizers?

Regards,

Julien

**Comment by mementum** [comment]
Hi,

`Sizers` as conceived for the platform have the ability to reach the `broker` and understand how much money (`cash`) is available. The code is fairly simple so I am quoting it:

```python
class SizerBase(with_metaclass(MetaParams, object)):

    params = (('broker', None,),)

    def getsizing(self, data, broker=None):
        broker = broker or self.params.broker
        return self._getsizing(broker.getcommissioninfo(data),
                               broker.getcash())

    def _getsizing(self, comminfo, cash):
        raise NotImplementedError

    def setbroker(self, broker):
        self.params.broker = broker

    def getbroker(self):
        return self.params.broker


class SizerFix(SizerBase):
    params = (('stake', 1),)

    def _getsizing(self, comminfo, cash):
        return self.params.stake

    def setsizing(self, stake):
        self.params.stake = stake
```

As you see there is a `broker` attribute (via the `params` or `p` attribute) which allows retrieving `cash` and of course any other thing held by the broker.

You only need to override the `_getsizing` method to create your own `Sizer`, which is already invoked with the available `cash`

And then in the `__init__` method of your strategy:

```python
class MyStrategy(bt.Strategy):
    def __init__(self):
        self.setsizer(MySizer())
```

When you issue a `buy` or `sell` the following happens:

```python
        size = size or self.getsizing(data)
```

So unless you specify a specific `size` in your call the `Sizer` instance present in the strategy will be finally called to retrieve the actual stake.

This seems to differ from things like `adddata`, `replaydata`, `addwriter` and other APIs which are all in the `Cerebro` class and yes it differs by design:

  - `backtrader` can run multiple strategies at the same time and it seems therefore reasonable to allow each strategy to have a different `Sizer`.

With regards to the plans about other `Sizers`: the idea is to build upon the concepts laid out by Van K. Tharp and build a couple of sizers.

If you have anything specific in mind and some literature, let me know.

**Comment by JulienNic0las** [comment]
Hello,

I fear I do not fully understand how work the sizer, the broker and strategy together. 

When the `buy()` method is called in the strategy without specified size the `self.getsizing(data)` method of the sizer is called which in turn will call the `_getsizing(self, comminfo, cash)` method. Ok. But data are not passed to this method. Thereby I do not see how the sizer can determine by its own the stakes to buy.

This leads me to try this:

_`
class MySizer(with_metaclass(MetaParams, object)):

    params = (
        ('broker', None,),
        ('stake', 1),
        )

    def getsizing(self, data, broker=None):
        broker = broker or self.params.broker
        return self._getsizing(broker.getcommissioninfo(data),
                               broker.getcash(), data=data)

    def _getsizing(self, comminfo, cash, data=None):
        if not data:
            return self.params.stake
            
        # Get total available cash and portfolio value
        cash = self.params.broker.getcash()
        value = self.params.broker.getvalue()
        # Get asset price to determine right number of shares to buy
        price = data.close[0]
        # Determine number of shares
        shares = cash / price
        return shares
        
    def setsizing(self):
        self.params.stake = stake

    def setbroker(self, broker):
        self.params.broker = broker

    def getbroker(self):
        return self.params.broker
`

With `MySizer` I understand that I'm not respecting the platform philosophy, but it is just a test. The `_getsizing` is provided with data and then can try to determine the number of stakes to return. But I fear this is strictly similar to compute stakes directly in the strategy class and call the `buy` method with the size as argument, isn't it? As a results, if I came to ask for a number of shares corresponding to all my current amount of cash and if the price rises, the order will not be accepted, right? 

At least this is what I currently have with a simple moving average strategy (if price up-cross the rolling mean buy, if cross down sell):
![figure_0-1](https://cloud.githubusercontent.com/assets/16461972/14403506/3bd853ea-fe59-11e5-81f0-f3e6a8e41c70.png)
As you can see I missed a buy. 

So, what I'm trying to do is to determine how many stakes to buy at the same _time_ the broker complete the order. I'm not fully understanding when the broker complete the order and when it talks to the sizer. 

Feel free to respond that I'm looking for something not realistic, and that I should work in another direction. I am still a neophyte in finance.

With regards,

Julien

**Comment by mementum** [comment]
Hi,

As stated in other ticket, the `Sizers` concept is still an underdeveloped one. Obviously not having the `data` (which is the target of the operation) as an argument of `_getsizing` is an oversight, which as done by you can be corrected by having `data=None` as a default argument.

It's not about *respecting* the platform. It is open to play, experiment, modify and adapt to what you need.

With regards to your questions:

  - Yes, the `sizer` operates as if you had calculated the stakes manually in the `strategy`. It is just an abstraction concept to encapsulate logic in a reusable manner.

  - The `broker` is a mere order taker and executor.

  - Yes, it you place an order with all your cash and the price raises your order will not be accepted (the 1st step is submission) if this is a "Market" or `Close`order.

  - The only way effectively control the price (and therefore the cash) is to use `Limit` orders (the broker has logic to try to give you a better price if possible by checking the ohlc prices to see if the limit can be improved)

The real question in the background seems to be the actual execution point of the order. First an axiom which is part of the design:

  - When the strategy's `next` method is called, the strategy evaluates a bar of prices. A bar which is already "closed", i.e: it has already happened. It's the past and the prices contained in the `high` - `low` range may not be seen again ever.

With that in mind the broker operates as follows for the order types:

  - `Market`

    The opening price of the next bar is used. The issues wants to get matched at the next tick and the next new tick is the next opening price.

  - `Close`

    This order type is meant for intraday backtesting. If the current development of the day events, then the order will be executed with the last "tick" of the session: the `close` price of the session. Still the order will first executed with the next bar. Of course if the order is when the last bar of day `x` is seen, the order will be executed with the last tick of day `x + 1`

  - `Limit`

    The order will be executed starting with the next bar if the limit price is seen. In some cases a better price will be attained, like if the `open` price is below the specified limit.
  
In your case

  - Your strategy`s `next` method is called at bar `x`

  - You issue an order for the entire amount of your cash.

  - The `broker` takes it and marks it as submitted.

  - The system fetches bar `x + 1`

  - The broker checks if the order can be `accepted` (enough cash) and if not issues a `margin` call and rejects the order

  - Order execution takes place (if possible and depending on type order)

  - Your strategy's `next` method is called again and you can evaluate the happenings at point `x + 1`

With this in mind it should be clear why your order is not being executed: not enough cash. (whether it is sensible to use all the cash is a topic for another discussion)

Is it unrealistic what you want? I cannot really answer this question.

But I can answer a similar one: is it realistic what `backtrader` does? That fully depends on the expectations of the end user. The real world is made of an unlimited number of cores and threads executing simultaneously and producing events (ticks, orders ...) and the platform operates with "past" data.

For me, the author, it is enough and I am happy with recreating a day via `replay` from `1 minute` bars. It's not the same thing I saw when looking at the live development of the chart but it is enough for me for all intents and purposes.

An analogy would be that of digital and analog recordings. The digital ones make a compromise to represent the analog reality in terms of sampling bits and sampling frequency. Some information is lost and even more when compressed to things like mp3, but the result is good enough when played back even it is not exactly the same as the original.

**Comment by JulienNic0las** [comment]
Thanks a lot for your response. I really appreciate. 

I have a better understanding of how things work in `backtrader`. To be honest I'm testing (to not say playing around) different things with your platform. That's why I've a lot of questions. 
My intent is not to perform any kind of 'professional' backtests but whether to verify by my own trading strategies I came to read (both checking their claims to be wining strategies and if those can work outside the US markets and for other assets than those usually presented in white papers). Thereby I can afford simplifications and work arounds if the main points are still considered (commissions and so on). As far as I can see, I can do it simply with you platform and that's great!

I made a simple work around with the initial problem which motivated me to open this ticket: I ask to buy X% of the total available cash with X as close as possible to 100. To determine X I can compute the highest positive daily return over the observed period before running the backtest (plus commission). With this method I'm sure to have always enough cash for the transaction.

Thanks for the commit.

J
